parlai/nn/lr_scheduler.py

```diff
@@ -0,0 +1,508 @@
+
+        This uses the loss as the validation metric if present, if not this
+        function does nothing. Note that the model must be reporting loss for
+        this to work.
+
+        Override this method to override the behavior for validation schedulers.
+        """
+        pass
+
+
+class ReduceOnPlateauLRScheduler(ParlAILRScheduler):
+    """
+    Scheduler that decays by a multiplicative rate when valid loss plateaus.
+    """
+
+    def __init__(
+        self,
+        optimizer,
+        states,
+        hard_reset,
+        patience,
+        decay,
+        warmup_updates,
+        warmup_rate,
+    ):
+        super().__init__(optimizer, states, hard_reset, warmup_updates, warmup_rate)
+        self.scheduler = optim.lr_scheduler.ReduceLROnPlateau(
+            optimizer, 'min', factor=decay, patience=patience, verbose=True
+        )
+
+    def train_step(self, scheduler_steps):
+        pass
+
+    def valid_step(self, metrics_dict):
+        if self._is_lr_warming_up():
+            # we're not done warming up, so don't start using validation
+            # metrics to adjust schedule
+            return
+        if 'loss' not in metrics_dict:
+            # nothing to step on, just skip
+            warn_once("LR scheduler expected to see loss metric, but didn't.")
+            return
+        self.scheduler.step(metrics_dict['loss'])
+
+
+class FixedLRScheduler(ParlAILRScheduler):
+    """
+    Scheduler that decays by a fixed multiplicative rate at each valid step.
+    """
+
+    def __init__(
+        self,
+        optimizer,
+        states,
+        hard_reset,
+        patience,
+        decay,
+        warmup_updates,
+        warmup_rate,
+    ):
+        super().__init__(optimizer, states, hard_reset, warmup_updates, warmup_rate)
+        self.scheduler = optim.lr_scheduler.StepLR(optimizer, patience, gamma=decay)
+
+    def train_step(self, scheduler_steps):
+        pass
+
+    def valid_step(self, metrics_dict):
+        if self._is_lr_warming_up():
+            # we're not done warming up, so don't start using validation
+            # metrics to adjust schedule
+            return
+        self.scheduler.step()
+
+
+class InvSqrtLRScheduler(ParlAILRScheduler):
+    """
+    Scheduler that decays at an inverse square root rate.
+    """
+
+    def __init__(
+        self,
+        optimizer,
+        states,
+        hard_reset,
+        patience,
+        decay,
+        warmup_updates,
+        warmup_rate,
+        invsqrt_lr_decay_gamma,
+    ):
+        """
+        invsqrt_lr_decay_gamma determines the cycle length of the inverse square root
+        scheduler.
+
+        When steps taken == invsqrt_lr_decay_gamma, the lr multiplier is 1
+        """
+        super().__init__(optimizer, states, hard_reset, warmup_updates, warmup_rate)
+        if invsqrt_lr_decay_gamma <= 0:
+            raise ValueError(
+                '--lr-scheduler invsqrt requires setting --invsqrt_lr_decay_gamma'
```

Comment:
please update error message to dashes, and --invsqrt-lr-decay-gamma should default to whatever warmup_updates was if it's not available, for backwards compatibility.

Reply:
(also the unit tests are coming from the lack of infering gamma from warmup as is the past behavior)